### DISCUSSION TITLE
Could you provide an example of pagination?

### DISCUSSION THREAD

**Discussion by maxixcom** [Netflix/dgs-framework#61] [Q&A]
Hello and thank you for your work. I'm new in GraphQL and the framework seems the easiest thing to dive in with. But I'm stuck with pagination. I understand that I can add a couple of parameters like `limit`, `offset` for this. But I don't understand where a meta information should be in response with info like `totalCount`.
I have found suggestions from here [GraphQL Pagination](https://graphql.org/learn/pagination/#pagination-and-edges) but it seems a lot of additional abstractions take place there like Edges, Nodes, Cursor, PageInfo etc. Do you have anything ready for that kind of pagination? Or could you suggest other library that  could help me with that?

**Comment by srinivasankavitha** [comment]
`graphql-java` supports relay spec based pagination. The `graphql.relay` package provides access to those types for you to use in your implementation. We don't have any additional conveniences built-into the framework around pagination currently.  You can define your schema, however, to support pagination and implement the datafetchers accordingly. We will add a full example implementation with pagination in the future. For now, here is an example schema that uses cursors for pagination based on the relay-spec (hence not complete): 

```
type Query @extends {
    ticks(first: Int, after: Int): TicksConnection
}

type Tick {
    id: ID
    route: Route
    userRating: String
    comments: String
}

type Route {
    routeId: ID
    name: String
    grade: String
    location: [String]
}

type TicksConnection {
    pageInfo: PageInfo!
    edges: [TickEdge]
}

type TickEdge {
    cursor: String!
    node: Tick
}
```

With this schema, you can implement your data fetcher like this:

```
@DgsData(parentType = "Query", field = "ticks")
    fun ticks(dataFetchingEnvironment: DataFetchingEnvironment): Connection<Tick> {

        val first = dataFetchingEnvironment.arguments["first"] as Int? ?: 10
        val after = dataFetchingEnvironment.arguments["after"] as Int? ?: 0
       
        val ticksResult = someAPI.loadTicks(after, first)
        val edges = ticksResult.ticks.mapIndexed { idx, tick -> DefaultEdge(tick, DefaultConnectionCursor(idx.toString())) }

        return DefaultConnection(edges, DefaultPageInfo(edges[0].cursor, edges[ticksResult.ticks.size - 1].cursor, after > 0, ticksResult.hasMore))
    }
```
Where `DefaultEdge`, `DefaultConnection`, `DefaultConnectionCursor` and `DefaultPageInfo` are provided by `graphql-java`

Hope this helps.

**Reply by maxixcom** [reply]
Thank you, it really helps!

**Reply by ferrisyang** [reply]
How to solve the error [The field type 'PageInfo' is not present] when start program?

**Reply by ferrisyang** [reply]
#371 I found one solution.
